### PR TITLE
Tractable bivariate distances

### PR DESCRIPTION
**Pull request by ships**
## Summary

Rewrite distance calculations using JAX for GPU-accelerated vectorized operations. Replaces scipy/numpy-based implementations with JIT-compiled JAX functions, achieving significant speedups for bivariate distance computations on large dataframes.

### Core: JAX-based Distance Calculations

**counting.py**
- Add `OrdinalDF` class: encodes categorical dataframes into integer JAX arrays using sklearn's OrdinalEncoder. Null values map to -1 sentinel. Shared encoders across dataframes ensure consistent category mapping.
- Add `normalize_count_memoized`: JIT-compiled counting via vectorized scatter-add (`jnp.zeros().at[idx].add()`). Replaces `collections.Counter` / `pl.Series.value_counts()`.
- Add `normalize_count_bivariate_memoized`: JIT-compiled co-occurrence counting for column pairs. Computes linear index for 2D scatter-add.
- Wrapper functions (`normalize_count`, `normalize_count_bivariate`) preserved for backward compatibility; delegate to memoized implementations.

**distances.py**
- Replace scipy entropy/jensenshannon with JAX implementations (`fasttvd`, `fastkl`, `fastjs`).
- `DistanceMetric` IntEnum enables `jax.lax.switch` for JIT-compatible metric selection.
- `univariate_distance` and `bivariate_distance` are now JIT-compiled, operating on pre-encoded integer arrays.
- `bivariate_distances_in_data`: groups column pairs by cardinality and uses `vmap` for batched distance computation.

### Build & Tooling
- Migrate to `uv` package manager (adds uv.lock)
- Add ruff for linting
- Add jax, jaxlib, equinox, jaxtyping dependencies
- Performance tests in `tests/test_distances_performance.py`, excluded from default pytest runs

### API
- `harmonize_categorical_probabilities` deprecated (use `OrdinalDF.from_dataframes()`)
- Distance functions now operate on encoded arrays; high-level `*_in_data` functions handle encoding automatically
- Empty input raises `IndexError` (sklearn validation) instead of `ValueError`

## Test Plan
- [x] All existing tests pass (112 passed)
- [x] Performance benchmarks verify speedup on bivariate/univariate distance calculations
- [x] Null handling verified across list and pl.Series inputs
- [x] New `test_ordinal_df.py` covers OrdinalDF encoding edge cases

### CODE REVIEW

File: tests/test_ordinal_df.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+
+def test_ordinal_df_single_column_two_values():
+    df = pl.DataFrame({"col1": ["a", "b", "a", "b"]})
+    odf = OrdinalDF.from_dataframe(df)
+
+    assert odf.columns == ("col1",)
+    assert len(odf.encoders) == 1
+    assert odf.data.shape == (4, 1)
+    assert set(odf.data[:, 0].tolist()) == {0, 1}
```
Comment (Schaechtle) [inline tests/test_ordinal_df.py:15]: There are plenty of assert testing shapes and unique values via `set`, but not a single one test against exactly the coding (I guess `[0, 1, 0, 1]` in this case). We should have that, no?

**Review by Schaechtle — APPROVED** [review]
thank you! minor comment/request for fix inline. Do not need to re-review.
